misc/pstring.cpp

```diff
@@ -38,140 +38,22 @@
  * $NoKeywords: $
  */
 
-#include <string.h>
-#include <stdarg.h>
-#include <stdio.h>
-#include <stdlib.h>
-#include "pstring.h"
-
-#ifdef __LINUX__
-#include "lnxfix.h"
-#endif
+#include <cstring>
 
-//	Pvsprintf
-//		Similar to vsprintf/_vsnprintf, however handles the case of a buffer overflow.  Arguments are the
-//	same as _vsnprintf.  In the case of count, pass in the size of the buffer, in the case where there is
-//	a buffer overflow, a \0 will be tacked on as the last byte in the buffer, ensuring a valid string.
-int Pvsprintf(char *buffer, int count, const char *format, va_list argptr) {
-  int ret = _vsnprintf(buffer, count - 1, format, argptr);
-  if (ret == -1)
-    buffer[count - 1] = '\0';
-  return ret;
-}
+#include <algorithm>
+#include <string>
 
-//	Psprintf
-//		Similar to sprintf/_snprintf, however handles the case of a buffer overflow.  Arguments are the
-//	same as _snprintf.  In the case of count, pass in the size of the buffer, in the case where there is
-//	a buffer overflow, a \0 will be tacked on as the last byte in the buffer, ensuring a valid string.
-int Psprintf(char *buffer, int count, const char *format, ...) {
-  va_list ap;
-  va_start(ap, format);
-  int ret = Pvsprintf(buffer, count, format, ap);
-  va_end(ap);
-  return ret;
-}
+#include "pstring.h"
 
 // CleanupStr
-//		this function strips all leading and trailing spaces, keeping internal spaces.  this goes
-//		for tabs too.
-int CleanupStr(char *dest, const char *src, int destlen) {
-  int i, j, err, begin = 0, end = 0, len;
-
-  err = 0;
-
-  len = strlen(src);
-  for (i = 0; i < len; i++) {
-    char ch;
-    ch = src[i];
-
-    //	 mark beginning.
-    if ((ch > ' ' && ch > '\t') && err < 1) {
-      err = 1;
-      begin = i;
-      end = i;
-    } else if (ch == ' ' && err == 1) {
-      err = 2;
-      end = i;
-    } else if (ch > ' ' && err >= 1) {
-      end = i;
-    }
+//    This function strips all leading and trailing spaces, keeping internal spaces. This goes for tabs too.
+int CleanupStr(char *dest, const char *src, std::size_t destlen) {
+  if ((src != nullptr) && (dest != nullptr)) {
+    std::string s{src};
+    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) { return !std::isspace(ch); }));
+    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) { return !std::isspace(ch); }).base(), s.end());
+    std::strncpy(dest, s.c_str(), destlen);
```

Comment:
Since you are benchmarking, a `std::string_view` and `remove_prefix/suffix` (since C++17) should be faster :)

Reply:
```cpp
int CleanupStr(char *dest, const char *src, std::size_t destlen) {
  if ((src != nullptr) && (dest != nullptr)) {
    std::string_view str{src};
    auto it1 = std::find_if_not(str.begin(), str.end(), [](char ch) { return std::isspace(ch); });
    str = str.substr(it1 - str.begin());

    auto it2 = std::find_if_not(str.rbegin(), str.rend(), [](char ch) { return std::isspace(ch); });
    str = str.substr(0, str.rend() - it2);

    int size = str.size() < destlen ? str.size() : destlen;
    std::strncpy(dest, str.data(), size);
    dest[size] = '\0';
  }
```

Would be faster, but no so fast as C:

```c
int CleanupStr(char *dest, const char *src, size_t destlen) {
  if (destlen == 0)
    return 0;

  const char *end;
  size_t out_size;

  // Trim leading space
  while (isspace((unsigned char)*src))
    src++;

  // All white spaces
  if (*src == 0) {
    *dest = 0;
    return 0;
  }

  // Trim trailing space
  end = src + strlen(src) - 1;
  while (end > src && isspace((unsigned char)*end))
    end--;
  end++;

  out_size = (end - src) < destlen - 1 ? (end - src) : destlen - 1;

  memcpy(dest, src, out_size);
  dest[out_size] = 0;

  return 0;
}
```
(Not tested properly, just need careful check)